packages/oidc-middleware/src/ExpressOIDC.js

```diff
@@ -68,28 +71,35 @@ module.exports = class ExpressOIDC extends EventEmitter {
     // Validate the client_secret param
     assertClientSecret(client_secret);
 
-    // Validate the redirect_uri param
-    assertRedirectUri(redirect_uri);
+    // Validate the appBaseUrl param
+    assertAppBaseUrl(appBaseUrl);
 
     // Add defaults to the options
-    options = _.merge({
+    options = merge({
       response_type: 'code',
       scope: 'openid',
       routes: {
         login: {
           path: '/login'
         },
-        callback: {
+        loginCallback: {
           path: '/authorization-code/callback',
-          defaultRedirect: '/'
+          afterCallback: '/'
         }
       },
-      sessionKey: sessionKey || `oidc:${options.issuer}`,
+      sessionKey: sessionKey || `oidc:${issuer}`,
       maxClockSkew: 120
     }, options)
 
+    // Build redirect uri unless explicitly set
+    options.loginRedirectUri = loginRedirectUri || `${appBaseUrl}${options.routes.loginCallback.path}`;
+
+    // Validate the redirect_uri param
+    assertRedirectUri(options.loginRedirectUri);
+
     const context = {
-      options
+      options,
+      emitter: this
```

Comment:
Curious, what's the reason for this change? To make it more explicit?

Reply:
Currently, only the ExpressOIDC object or something with a reference to it can emit.  

Now that we can have errors on more than setup, (such as an error during token revoke) we need a way to emit errors on those, and those don't happen in ExpressOIDC, but rather in another call that ExpressOIDC calls.  

We could have THOSE be emitters and have ExpressOIDC listen and re-emit, or we can just pass the emitter down so those can call the emitter directly when needed.